### PR TITLE
Gpg 738 set up ga tracking for filters

### PR DESCRIPTION
**Pull request by Lucas-Fernandes-Nogueira**
T - I've used the Google Analytics Debugger Chrome extension to check if the event were being sent and their properties and it all look good.
R - Low risk, since it doesn't affect the user. 
D - I'll add a page on Swiki about the GA setup.

### CODE REVIEW

**Comment by Lucas-Fernandes-Nogueira** [comment]
There's two commits:

- The first contains the actual GA tracking implementation.
- The second removes old unnecessary/ commented out code.

File: GenderPayGap.WebUI/Scripts/Application/LiveSearch.js [hunk 1]
```diff
@@ -94,26 +95,23 @@
...
-                liveSearch.displayResults(response, this.searchState);
-                liveSearch.clearLoadingIndicator();
-            }).error(function () {
-                window.location = '/error/1146';
-            });
-        //} else {
-            //this.displayResults(cachedResultData, searchState);
-            //var out = new $.Deferred()
-            //return out.resolve();
-        //}
```
Comment (Lucas-Fernandes-Nogueira) [inline GenderPayGap.WebUI/Scripts/Application/LiveSearch.js:116]: Removed commented out code.

File: GenderPayGap.WebUI/Scripts/Application/LiveSearch.js [hunk 2]
```diff
@@ -94,26 +95,23 @@
 
     LiveSearch.prototype.updateResults = function updateResults() {
         var searchState = $.param(this.state);
-        //var cachedResultData = this.cache(searchState);
```
Comment (Lucas-Fernandes-Nogueira) [inline GenderPayGap.WebUI/Scripts/Application/LiveSearch.js:97]: Removed commented out code.

File: GenderPayGap.WebUI/Scripts/Application/LiveSearch.js [hunk 3]
```diff
@@ -94,26 +95,23 @@
 
     LiveSearch.prototype.updateResults = function updateResults() {
         var searchState = $.param(this.state);
-        //var cachedResultData = this.cache(searchState);
         var liveSearch = this;
-        //if (typeof cachedResultData === 'undefined') {
```
Comment (Lucas-Fernandes-Nogueira) [inline GenderPayGap.WebUI/Scripts/Application/LiveSearch.js:99]: Removed commented out code.

File: GenderPayGap.WebUI/Scripts/Application/LiveSearch.js [hunk 4]
```diff
@@ -132,9 +130,6 @@
         if (action == $.param(this.state)) {
             this.clearLoadingIndicator();
             this.$resultsBlock.html(results);
-            //this.$resultsBlock.mustache('finders/_results', results);
-            //this.$loadingBlock.mustache('finders/_result_count', results);
-            //this.$atomAutodiscoveryLink.attr('href', results.atom_url);
```
Comment (Lucas-Fernandes-Nogueira) [inline GenderPayGap.WebUI/Scripts/Application/LiveSearch.js:137]: Removed commented out code.

File: GenderPayGap.WebUI/Scripts/Application/LiveSearch.js [hunk 5]
```diff
@@ -202,6 +197,65 @@
...
+                filter => filter.Group
+            ).includes(parameter.name)
+        );
+        let GAEvent = {
+            hitType: 'event',
+            eventCategory: this.GATrackFilters.category,
+            eventAction: [],
+            eventLabel: {}
+        };
+        sendGpgEvent(this.convertGAEventToHumanReadableFormat(this.addFiltersToGAEvent(selectedFilters, GAEvent)));
```
Comment (Lucas-Fernandes-Nogueira) [inline GenderPayGap.WebUI/Scripts/Application/LiveSearch.js:213]: sendGpgEvent is the function I defined in _Tracking.cshtml.

File: GenderPayGap.WebUI/Views/Viewing/Finder/SearchResults.cshtml [hunk 6]
```diff
@@ -53,17 +53,20 @@
             //Keep comparison basket within viewport
             GOVUK.stickAtTopWhenScrolling.init();
 
-            // Instantiate an option select for each one found on the page
-            var filters = $('#FinderForm .govuk-option-select').map(function () {
-                return new GOVUK.OptionSelect({ $el: $(this) });
-            });
-
```
Comment (Lucas-Fernandes-Nogueira) [inline GenderPayGap.WebUI/Views/Viewing/Finder/SearchResults.cshtml:60]: This code didn't seem to be used anywhere, at least the filters variable isn't.

File: GenderPayGap.WebUI/Views/Viewing/Finder/SearchResults.cshtml [hunk 7]
```diff
@@ -53,17 +53,20 @@
...
 
-            // Instantiate an option select for each one found on the page
-            var filters = $('#FinderForm .govuk-option-select').map(function () {
-                return new GOVUK.OptionSelect({ $el: $(this) });
-            });
-
             new GOVUK.LiveSearch({
                 formId: "FinderForm",
                 $results: $('.js-live-search-results-block'),
-                //$atomAutodiscoveryLink: $atomAutodiscoveryLink,
```
Comment (Lucas-Fernandes-Nogueira) [inline GenderPayGap.WebUI/Views/Viewing/Finder/SearchResults.cshtml:64]: Removed commented out code.

File: GenderPayGap.WebUI/Views/Viewing/Finder/SearchResults.cshtml [hunk 8]
```diff
@@ -53,17 +53,20 @@
...
                     $("#AddAllOrgsToCompare").toggle();
+                },
+                GATrackFilters: {
+                    category: "Search filters",
+                    filters: [
+                        @Html.Raw(Json.Serialize(Model.SizeFilterGroup)), 
+                        @Html.Raw(Json.Serialize(Model.SectorFilterGroup)), 
+                        @Html.Raw(Json.Serialize(Model.ReportingYearFilterGroup)), 
+                        @Html.Raw(Json.Serialize(Model.ReportingStatusFilterGroup))
+                        ]
```
Comment (Lucas-Fernandes-Nogueira) [inline GenderPayGap.WebUI/Views/Viewing/Finder/SearchResults.cshtml:69]: Any better way to write this?

**Comment by Lucas-Fernandes-Nogueira** [comment]
Examples Events sent to GA:

- **eventCategory**: Search filters
- **eventAction**: Reporting year; Reporting status
- **eventLabel**: Reporting year: 2019 to 2020; Reporting status: Reported with extra information; 

- **eventCategory**: Search filters
- **eventAction**: Employer size; Sector; Reporting year; Reporting status
- **eventLabel**: Employer size: 1000 to 4999, 20,000 or more; Sector: Administrative and support service activities, Agriculture Forestry and Fishing; Reporting year: 2019 to 2020; Reporting status: Reported with extra information;

**Comment by Lucas-Fernandes-Nogueira** [comment]
When no filters are selected, we still send an event with the following format:
- eventCategory: Search filters
- eventAction:
- eventLabel:
Would it be better if we don't send this at all?

**Comment by Lucas-Fernandes-Nogueira** [comment]
> When no filters are selected, we still send an event with the following format:
> 
> * eventCategory: Search filters
> * eventAction:
> * eventLabel:
>   Would it be better if we don't send this at all?

Actually, it seems like since eventAction is a required field, then we're not sending these events. I also checked the GA and can't see any of these events. There's nothing in the code that stops these events from being sent, so I assume the ga("send", Event) line automatically fails to send events if action is empty.
Should I still make sure we don't reach this line if action is empty?

File: GenderPayGap.WebUI/Scripts/Application/LiveSearch.js [hunk 9]
```diff
@@ -202,6 +197,65 @@
     LiveSearch.prototype.collapseFilters = function collapseFilters() {
         $(".govuk-option-select.js-collapsible .js-container-head[aria-expanded=true]").click();
     };
+    
+    LiveSearch.prototype.sendFilterEventToGA = function sendFilterEventToGA() {
+        let selectedFilters = this.state.filter(
+            parameter =>  this.GATrackFilters.filters.map(
+                filter => filter.Group
+            ).includes(parameter.name)
```
Comment (malinaslevoaca1) [inline GenderPayGap.WebUI/Scripts/Application/LiveSearch.js:205]: We can use [some](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some) instead.
```
this.GATrackFilters.filters.some(filter => filter.Group === parameter.name)
```

File: GenderPayGap.WebUI/Scripts/Application/LiveSearch.js [hunk 10]
```diff
@@ -202,6 +197,65 @@
...
+    LiveSearch.prototype.convertEventLabelToHumanReadableString = function convertEventLabelToHumanReadableString(eventLabel) {
+        let eventLabelString = '';
+        for (const filterGroup in eventLabel)
+        {
+            eventLabelString += this.GATrackFilters.filters.find(filter => filter.Group === filterGroup).Label + ': ';
+            eventLabelString += eventLabel[filterGroup].map(
+                value => this.GATrackFilters.filters.find(
+                    filter => filter.Group === filterGroup).Metadata.find(
+                        filter => filter.Value === value).Label
+            ).join(', ');
```
Comment (malinaslevoaca1) [inline GenderPayGap.WebUI/Scripts/Application/LiveSearch.js:254]: I suggest we move `this.GATrackFilters.filters.find(filter => filter.Group === filterGroup)` to a separate variable. 
I find this quite hard to read. Can you please add some comments? 

Maybe we can build the object we need when iterating through the arrays for the first time in order to have here all the labels. What do you think?

**Review by malinaslevoaca1 — COMMENTED** [review]
Nice! Some suggestions:

File: GenderPayGap.WebUI/Scripts/Application/LiveSearch.js [hunk 11]
```diff
@@ -202,6 +197,65 @@
...
     };
+    
+    LiveSearch.prototype.sendFilterEventToGA = function sendFilterEventToGA() {
+        let selectedFilters = this.state.filter(
+            parameter =>  this.GATrackFilters.filters.map(
+                filter => filter.Group
+            ).includes(parameter.name)
+        );
+        let GAEvent = {
+            hitType: 'event',
```
[the remark this answers is not in the document]
Comment (Lucas-Fernandes-Nogueira) [reply GenderPayGap.WebUI/Scripts/Application/LiveSearch.js:208]: Agree!

File: GenderPayGap.WebUI/Scripts/Application/LiveSearch.js (hunk 9 above)
Comment (Lucas-Fernandes-Nogueira) [reply GenderPayGap.WebUI/Scripts/Application/LiveSearch.js:205]: That's neat, I didn't know about this! Thanks :)

File: GenderPayGap.WebUI/Scripts/Application/LiveSearch.js (hunk 10 above)
Comment (Lucas-Fernandes-Nogueira) [reply GenderPayGap.WebUI/Scripts/Application/LiveSearch.js:254]: I agree it might be easier to do it all in one go. I'll try this :).

Comment (Lucas-Fernandes-Nogueira) [reply GenderPayGap.WebUI/Scripts/Application/LiveSearch.js]: I've completely refactored the filter event creation code. I think it's way more readable now. Please let me know what you think 👍.

File: GenderPayGap.WebUI/Scripts/Application/LiveSearch.js [hunk 12]
```diff
@@ -202,6 +198,56 @@
...
+    };
+    
+    LiveSearch.prototype.getSelectedFiltersInGroup = function getSelectedFiltersInGroup(filterGroup) {
+        // the state has various parameters, some of which will be filters.
+        // this function returns an array with the parameters that are filters of the given filter group.
+        return this.state.filter(parameter => parameter.name === filterGroup.Group);
+    };
+    
+    LiveSearch.prototype.getFilterLabelFromGroup = function getFilterLabelFromGroup(filter, filterGroup) {
+        return  filterGroup.Metadata.find(
```
Comment (malinaslevoaca1) [inline GenderPayGap.WebUI/Scripts/Application/LiveSearch.js:235]: Minor nitpick - there are 2 spaces here

**Review by malinaslevoaca1 — DISMISSED** [review]
Nice!

File: GenderPayGap.WebUI/Scripts/Application/LiveSearch.js (hunk 10 above)
Comment (malinaslevoaca1) [reply GenderPayGap.WebUI/Scripts/Application/LiveSearch.js]: It is indeed, thank you!

**Review by malinaslevoaca1 — APPROVED** [review]

**Pull request merged**